File: lib/delphi/src/Thrift.Configuration.pas
(*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements. See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership. The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License. You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations
 * under the License.
 *)

unit Thrift.Configuration;

interface

uses
  SysUtils, Generics.Collections, Generics.Defaults;

const
  DEFAULT_RECURSION_LIMIT  = 64;
  DEFAULT_MAX_MESSAGE_SIZE = 100 * 1024 * 1024; // 100 MB
  DEFAULT_MAX_FRAME_SIZE   = 16384000;      // this value is used consistently across all Thrift libraries

  DEFAULT_THRIFT_TIMEOUT = 5 * 1000; // ms

type
  IThriftConfiguration = interface
    ['{666F7848-744A-4746-BDD5-43DC9B1D5520}']
    function  GetRecursionLimit : Integer;
    procedure SetRecursionLimit( const value : Integer);
    function  GetMaxFrameSize : Integer;
    procedure SetMaxFrameSize( const value : Integer);
    function  GetMaxMessageSize : Integer;
    procedure SetMaxMessageSize( const value : Integer);

    property  RecursionLimit : Integer read GetRecursionLimit write SetRecursionLimit;
    property  MaxFrameSize : Integer   read GetMaxFrameSize   write SetMaxFrameSize;
    property  MaxMessageSize : Integer read GetMaxMessageSize write SetMaxMessageSize;
  end;


  TThriftConfigurationImpl = class( TInterfacedObject, IThriftConfiguration)
  strict private
    class procedure ValidateLimitArgument( const value : Integer); inline;

  strict protected
    FRecursionLimit : Cardinal;
    FMaxFrameSize   : Cardinal;
    FMaxMessageSize : Cardinal;

    // IThriftConfiguration
    function  GetRecursionLimit : Integer;
    procedure SetRecursionLimit( const value : Integer);
    function  GetMaxFrameSize : Integer;
    procedure SetMaxFrameSize( const value : Integer);
    function  GetMaxMessageSize : Integer;
    procedure SetMaxMessageSize( const value : Integer);

  public
    constructor Create;
  end;


implementation


{ TThriftConfigurationImpl }


constructor TThriftConfigurationImpl.Create;
begin
  inherited Create;

  FRecursionLimit := DEFAULT_RECURSION_LIMIT;
  FMaxFrameSize   := DEFAULT_MAX_FRAME_SIZE;
  FMaxMessageSize := DEFAULT_MAX_MESSAGE_SIZE;
end;


class procedure TThriftConfigurationImpl.ValidateLimitArgument( const value : Integer);
begin
  if value <= 0  // zero makes not much sense either
  then raise EArgumentOutOfRangeException.Create('Value must be positive');
end;


function TThriftConfigurationImpl.GetRecursionLimit: Integer;
begin
  result := FRecursionLimit and MAXINT;
  ASSERT( result > 0);
end;


procedure TThriftConfigurationImpl.SetRecursionLimit(const value: Integer);
begin
  ValidateLimitArgument( value);
  ASSERT( value > 0);
  FRecursionLimit := value and MAXINT;
end;


function TThriftConfigurationImpl.GetMaxFrameSize: Integer;
begin
  result := FMaxFrameSize and MAXINT;
  ASSERT( result > 0);
end;


procedure TThriftConfigurationImpl.SetMaxFrameSize(const value: Integer);
begin
  ValidateLimitArgument( value);
  ASSERT( value > 0);
  FMaxFrameSize := value and MAXINT;
end;


function TThriftConfigurationImpl.GetMaxMessageSize: Integer;
begin
  result := FMaxMessageSize and MAXINT;
  ASSERT( result > 0);
end;


procedure TThriftConfigurationImpl.SetMaxMessageSize(const value: Integer);
begin
  ValidateLimitArgument( value);
  ASSERT( value > 0);
  FMaxMessageSize := value and MAXINT;
end;


end.

File: lib/c_glib/src/thrift/c_glib/thrift_configuration.c
/*
 *  Licensed to the Apache Software Foundation (ASF) under one
 *  or more contributor license agreements. See the NOTICE file
 *  distributed with this work for additional information
 *  regarding copyright ownership. The ASF licenses this file
 *  to you under the Apache License, Version 2.0(the
 *  "License"); you may not use this file except in compliance
 *  with the License. You may obtain a copy of the License at
 *    
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing,
 *  software distributed under the License is distuributed on an
 *  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY 
 *  KIND, either express or implied. See the License for the
 *  specific language governing permissions and limitations
 *  under the License.
 */

#include <thrift/c_glib/thrift.h>
#include "thrift_configuration.h"

/* object properties */
enum _ThriftConfigurationProperties
{
  PROP_0,
  PROP_THRIFT_CONFIGURATION_MAX_MESSAGE_SIZE,
  PROP_THRIFT_CONFIGURATION_MAX_FRAME_SIZE,
  PROP_THRIFT_CONFIGURATION_RECURSION_LIMIT
};

G_DEFINE_TYPE(ThriftConfiguration, thrift_configuration, G_TYPE_OBJECT)

/* property accessor */
void
thrift_configuration_get_property(GObject *object, guint property_id,
		                   GValue *value, GParamSpec *pspec)
{
  ThriftConfiguration *configuration = THRIFT_CONFIGURATION(object);

  THRIFT_UNUSED_VAR (pspec);

  switch (property_id)
  {
    case PROP_THRIFT_CONFIGURATION_MAX_MESSAGE_SIZE:
         g_value_set_int(value, configuration->maxMessageSize_);
	 break;
    case PROP_THRIFT_CONFIGURATION_MAX_FRAME_SIZE:
         g_value_set_int(value, configuration->maxFrameSize_);
         break;
    case PROP_THRIFT_CONFIGURATION_RECURSION_LIMIT:
	 g_value_set_int(value, configuration->recursionLimit_);
	 break;
   }
}

/* property mutator */
void
thrift_configuration_set_property(GObject *object, guint property_id, 
		                  const GValue *value, GParamSpec *pspec)
{
  ThriftConfiguration *configuration = THRIFT_CONFIGURATION (object);

  THRIFT_UNUSED_VAR (pspec);

  switch (property_id)
  {
    case PROP_THRIFT_CONFIGURATION_MAX_MESSAGE_SIZE:
	 configuration->maxMessageSize_ = g_value_get_int (value);
	 break;
    case PROP_THRIFT_CONFIGURATION_MAX_FRAME_SIZE:
	 configuration->maxFrameSize_ = g_value_get_int (value);
	 break;
    case PROP_THRIFT_CONFIGURATION_RECURSION_LIMIT:
	 configuration->recursionLimit_ = g_value_get_int (value);
	 break;
  }
}
	
static void
thrift_configuration_class_init (ThriftConfigurationClass *cls)
{
  GObjectClass *gobject_class = G_OBJECT_CLASS (cls);
  GParamSpec *param_spec = NULL;

  /* setup accessors and mutators */
  gobject_class->get_property = thrift_configuration_get_property;
  gobject_class->set_property = thrift_configuration_set_property;

  param_spec = g_param_spec_int ("max_message_size",
                                 "max_message_size (construct)",
                                 "Set the max size of the message",
                                 0, /* min */
                                 G_MAXINT32, /* max */
                                 DEFAULT_MAX_MESSAGE_SIZE, /* default by convention */
                                 G_PARAM_CONSTRUCT_ONLY |
                                 G_PARAM_READWRITE);

  g_object_class_install_property (gobject_class, PROP_THRIFT_CONFIGURATION_MAX_MESSAGE_SIZE,
                                   param_spec);

  param_spec = g_param_spec_int ("max_frame_size",
                                 "max_frame_size (construct)",
                                 "Set the max size of the frame",
                                 0, /* min */                         
                                 G_MAXINT32, /* max */
                                 DEFAULT_MAX_FRAME_SIZE, /* default by convention */
                                 G_PARAM_CONSTRUCT_ONLY |
                                 G_PARAM_READWRITE);

  g_object_class_install_property (gobject_class, PROP_THRIFT_CONFIGURATION_MAX_FRAME_SIZE,
                                   param_spec);

  param_spec = g_param_spec_int ("recursion_limit",
                                 "recursion_limit (construct)",
                                 "Set the limit of the resursion",
                                 0, /* min */
                                 G_MAXINT32, /* max */
                                 DEFAULT_RECURSION_DEPTH, /* default by convention */
                                 G_PARAM_CONSTRUCT_ONLY |
                                 G_PARAM_READWRITE);

  g_object_class_install_property (gobject_class, PROP_THRIFT_CONFIGURATION_RECURSION_LIMIT,
                                   param_spec);

}

static void
thrift_configuration_init (ThriftConfiguration *configuration)
{
  THRIFT_UNUSED_VAR (configuration);
}
